designs/autoscaler.md

```diff
@@ -0,0 +1,515 @@
+
+This document aims to solve this limitation and provides **autoscaling** **compatibility** in EKS Anywhere, so the user can either choose to install Cluster Autoscaler as an EKS-A curated package during or after cluster creation time, or install themselves from [upstream](https://github.com/kubernetes/autoscaler/releases) after a cluster is created.
+
+### Goals and Objectives
+
+As an EKS Anywhere user, I want to:
+
+* create clusters in a streamlined fashion with autoscaling capabilities
+* add autoscaling capabilities to existing EKS-A clusters
+* see my clusters to auto scale up and down worker nodes within the boundaries I limited
+
+### Statement of Scope
+
+#### In Scope
+
+* Include Cluster Autoscaler as an [EKS-A curated package](https://anywhere.eks.amazonaws.com/docs/concepts/packages/).
+* Integrate Cluster Autoscaler to all CAPI providers since the feature is provider agnostic.
+* Integrate autoscaling configuration in cluster spec to make EKS-A cluster autoscaling-aware.
+* Generate an EKS-A opinionated Cluster Autoscaler curated package template with default configurations.
+* Install Cluster Autoscaler controller as a curated package during or after management/workload cluster creation.
+* Support upstream k8s Autoscaler installed/managed outside of EKS-A Curated Package.
+* Update the autoscaling configuration to set up the min/max number of nodes in each worker node group.
+
+#### Out of Scope
+
+* Compare cluster autoscaler products (e.g. [Karpenter](https://karpenter.sh/))
+* Autoscale at the pod level
+* Control Plane nodes autoscaling
+
+#### Future Scope
+
+* Install Cluster Autoscaler controller as a curated package during workload cluster creation (this feature not yet supported by EKS-A packages).
+* Support other autoscaling solutions (e.g. [Karpenter](https://karpenter.sh/)).
+
+## Overview of Solution
+
+We provide a solution to add autoscaling configuration in the EKS Anywhere API, which marks the worker node groups to be Cluster Autoscaler compatible. EKS Anywhere user can then install Cluster Autoscaler in any EKS-A cluster with proper configuration, and expect node autoscaling to work for the associated resource group.
+
+1. Introduce an optional type `AutoScalingConfiguration` in the cluster and populate it into each worker node group configuration.
+2. If `AutoScalingConfiguration` is specified (not nil) for a worker node group, EKS-A converts its fields to CAPI autoscaler annotations and applies them to the underlying CAPI Machine Deployment resource. In this way, the specified worker node group will be discoverable by the Cluster Autoscaler controller once installed.
+3. Install Cluster Autoscaler either (A) as an EKS-A curated package, or (B) from upstream; with configuration that uses Cluster API as cloud provider, and auto-discovers node groups in the target cluster.
+
+### API Design
+
```

Comment:
I would expect these to be configurations in the package custom resource.

Reply:
I agree, at least at first thought. Packages shouldn't be reliant on the cluster controller to perform their functions, at least not in terms of configuration. It wouldn't support the concept of packages on a non-eks-a cluster, for example, or it would make for multiple avenues of obtaining that information would have to be available.

Reply:
If the cluster controller wants to be sole point of access for anything in the domain of cluster api, I think it will need to be flexible in allowing curated packages access to at least parts of that functionality. It will be a major detriment to curated packages independent development and release cycles if they require custom coding in the cluster controller for every package that needs to touch something in the cluster controller's domain.